### ISSUE TITLE
Appending or "stacking" variable lists

### ISSUE THREAD

**Issue by rahniwalden** [ansible/ansible#17714]
##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

Ansible support for appending or "stacking" Variable Lists

In this case, with "with_items"
--Should be able to append variable lists on the fly, conditionally
##### ANSIBLE VERSION

```
ansible --version
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY

We have the goal of dynamically being able to append lists to another list in a variable during a task. This setup is for the configuration of sec on rhel 6. We have different variables we want appended to a base list, based on conditions being met by the host.

For example, we have three variables defined at the moment. One is listed in /defaults/main.yml. A basic list, then two others that correspond to additional configurations needed by different types of application hosts.

Let's call the vars:
base
app1 
app2

Pure python can handle recursion. If we set:

```
  base = base + app1
  then set: base = base + app2
  we get: base = base + app1 + app2
```

But Ansible does not appear to support this. It fails with a, "recursion detected in template..." error. If you fix the self-reference, Ansible  of course will only accepts the last value set, for example:

```
  base = config + app1
  then set: base = config + app2
  we get: base = config + app2
```

I've seen the same error pretty much spelled out here: https://groups.google.com/forum/#!topic/ansible-project/Jo50nu3g2b4

As of right now, we can get around this by listing all the variables (and only including the variables based on conditions being met by the host). For example:

```
- name: write /etc/sysconfig/sec 
  lineinfile: 
    state: present 
    dest: /path/to/file
    backup: yes 
    insertafter: EOF 
    line: "{{ item }}" 
  with_items: 
    - "{{ base }}"
    - "{{ app1 | default(omit) }}"
    - "{{ app2 | default(omit) }}"
```

We are required to continue appending variables under with_items, since currently, Ansible only uses the last definition of the variable.

But it would be nice to have ansible be able to handle appending variable values, so we could simply include something akin to the following in vars/app1.yml:

```

---
app1:  <value>
base: "{{ base + app1 }}"

```

Which would make it easier to manage large numbers of hosts which may need various mixes of settings applied. Including recursion in the variable file would cover all cases (appending the needed value to the base variable on the fly), instead of having to create the var.yml file and then still go back in and add the "{{ app# | default(omit) }}" to the list.....

**Comment by veksh** [comment]
Something like that would be very useful for extensible roles.

My current use case is generating a list of filters for syslog-ng.
In template I define a filter to exclude some messages from log, like this:

      ## high-volume app messages to exclude from catch-all local log ('messages')
      filter f_drop_local {
        filter(f_dummy)
      {% for filter in syslogd_local_filters %}
        or filter(f_{{ filter['name'] }})
      {% endfor %}
      ;};
      filter f_pass_local { not filter(f_drop_local); };

      ## local: all messages except too noisy f_drop_local
      destination messages {
        file("/var/log/messages"
        );
      };
      log { source(src); filter(f_pass_local); destination(messages); };

Set of filters is application-specific, and it would be just great if several groups were able to append to `syslogd_local_filters` in group vars like

      log_filters_ldap:
        - name: ldap_general_noise
          desc: "general openldap noisy events"
          expr: |
            facility(local5) and program("slapd") and level(debug)
            and (match('closed \(.*\)$')
              # redundant: always after BIND ... method=128
              or match('BIND dn=".*" mech=SIMPLE')
              or match('RESULT tag=97 err=0 text=')
              or match('UNBIND')
            )

      # error is "recursive loop detected in template string"
      syslogd_local_filters: "{{ syslogd_local_filters + log_filters_ldap }}"

For now, it is impossible to use generic syslog template when host is included in several
groups ("ldap_servers" and "mail_servers" have separate set of noise filters).
Usual work-around of assembling config from pieces with `assemble` does not apply
here too, so only way to generate config is to have explicit list of possible filter sets,
like mentioned above.

**Comment by linsomniac** [comment]
In the tasks you can do this using "set_fact" to set or override the fact with some new data, for example:

    - name: Append string to list
      set_fact:
        my_strings: "{{ my_strings }} + [ '{{ my_name }}' ]"

But, what I'd like to see is a way to extend vars, so that I can collect things from across host_vars, group_vars, and defaults, and run them in a base task.  For example:

    group_vars/all:
        apache_env_vars:
          - @ANSIBLE_APPEND@
          - APP_WEBROOT=/opt/app/web
        installed_packages:
          - @ANSIBLE_APPEND@
          - ssh
          - ack
        iptables_tcp:
          - @ANSIBLE_APPEND@
          - 10.0.0.0/8:22
    group_vars/webl
        apache_env_vars:
          - @ANSIBLE_APPEND@
          - WORKERS=200
        installed_packages:
          - @ANSIBLE_APPEND@
          - apache2
          - php
        iptables_tcp:
          - 80
          - 443

**Comment by kluner** [comment]
I would love to see this stacking of variables too.

My use case is that I have a bunch of FreeBSD servers, or Linux servers, that have very common configurations, like ACL's, but specific subsets have extra allowed hosts in the ACL's.

By simply stacking the variables, or appending them, this could be done in very reusable bit of playbook.

**Comment by carwyn** [comment]
This is a very common use case, especially when a config item is a list in a single file that needs appending to such as allowed users or groups in SSSD.

Being able to compose role default, host and group bars in this way would simplify things considerably.

**Comment by carwyn** [comment]
Note that a similar feature in LCFG (declaratice config tool from the 90s) had specific options to prefix, suffix or replace variables. It may have had type sensitive operators too (e.g. numericL addition, subtraction). It was a pretty fundamental feature and use cases well proven.

**Comment by bcoca** [comment]
fixed by #57894

**Comment by veksh** [comment]
I do not see how this improvement to `combine` will help here: idea was to keep some shared state variable and being able to append values to it, not to improve recursive list merge behavior.
